pulser/waveforms.py

```diff
@@ -279,24 +282,41 @@ def __getitem__(self, index_or_slice:
             array: np.ndarray = np.array([])
             wf_start = 0
             for wf in self._waveforms:
-                if wf_start <= start and start <= wf_start+wf.duration:
-                    if wf_start <= stop and stop <= wf_start+wf.duration:
+                duration: int = wf.duration
+                if start > wf_start + duration:
+                    # Skip this waveform, which is before the start index
+                    wf_start += duration
+                    continue
+                elif wf_start <= start:
+                    # The start index belongs to this waveform
+                    if wf_start <= stop <= wf_start+duration:
+                        # The stop index also belongs to this waveform
+                        # Hence, we can return a subset of its samples
                         return cast(np.ndarray,
-                                    wf.samples[start-wf_start:stop-wf_start])
+                                    wf[start-wf_start:stop-wf_start])
                     else:
+                        # Copy the part of the waveform after the start index
                         array = wf.samples[start-wf_start:]
-                if start < wf_start and stop > wf_start+wf.duration:
-                    array = np.concatenate((array, wf.samples))
-                if wf_start <= stop and stop <= wf_start+wf.duration:
-                    array = np.concatenate((array, wf.samples[:stop-wf_start]))
-                wf_start += wf.duration
+                elif start < wf_start:
```

Comment:
You can omit the comparison, it's the only option at this point

Reply:
👍